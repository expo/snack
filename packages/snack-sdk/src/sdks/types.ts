/**
 * Version of the sdk to use (e.g. "42.0.0").
 */
export type SDKVersion = '39.0.0' | '40.0.0' | '41.0.0' | '42.0.0' | '43.0.0';

/** @internal */
export type SDKSpec = {
  // Version-spec for the published "expo" package. This version is
  // used to fetch compatible package versions. The value is typically
  // in the form of "^39.0.0" but may also contain custom specs, such
  // as "40.0.0-beta.2".
  version: string;

  // Modules that are pre-loaded by the Snack runtime, and which
  // the user does not need to add to `package.json`.
  // The version for these packages is hardcoded as it cannot be
  // resolved using bundledNativeModules.json.
  coreModules: {
    [name: string]: string;
  };

  // Modules that are pre-loaded by the Snack runtime, but which
  // are expected to be added as a dependency anyway. This ensures
  // that `package.json` is up to date, when downloading the Snack
  // as a zip-file. These modules don't have versions defined, but
  // instead bundledNativeModules.json should be used to obtain
  // the recent version.
  bundledModules: {
    [name: string]: '*';
  };

  // Modules that have been deprecated. The value represents a valid
  // a description providing a solution or explanation (eg.
  // "expo-permissions": "Use permissions getters and requesters in specific modules instead...""
  deprecatedModules?: {
    [name: string]: string;
  };
};

/**
 * Feature that is supported by the SDK (e.g. "TYPESCRIPT").
 */
export type SDKFeature =
  | 'MULTIPLE_FILES'
  | 'PROJECT_DEPENDENCIES'
  | 'TYPESCRIPT'
  | 'UNIMODULE_IMPORTS'
  | 'POSTMESSAGE_TRANSPORT'
  | 'VERSIONED_SNACKAGER';
