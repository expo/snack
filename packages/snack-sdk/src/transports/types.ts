import { SnackFiles, SnackDependencies, SDKVersion } from 'snack-content';

import { ProtocolOutgoingMessage, ProtocolIncomingMessage } from './Protocol';

export type SnackTransportStartMessage = {
  type: 'start';
};

export type SnackTransportStopMessage = {
  type: 'stop';
};

export type SnackTransportUpdateCodeMessage = {
  type: 'update_code';
  data: {
    files: SnackFiles;
    dependencies: SnackDependencies;
    sdkVersion: SDKVersion;
  };
};

export type SnackTransportProtocolMessage = {
  type: 'protocol_message';
  data: ProtocolOutgoingMessage;
};

export type SnackTransportSyntheticMessage = {
  type: 'synthetic_event';
  data: any;
};

export type SnackTransportMessage =
  | SnackTransportStartMessage
  | SnackTransportStopMessage
  | SnackTransportUpdateCodeMessage
  | SnackTransportProtocolMessage
  | SnackTransportSyntheticMessage;

export type SnackTransportEventConnect = {
  type: 'connect';
  connectionId: string;
  data: any; // DeviceInfo
};

export type SnackTransportEventDisconnect = {
  type: 'disconnect';
  connectionId: string;
  data: {
    timedout: boolean;
  };
};

export type SnackTransportEventProtocolMessage = {
  type: 'protocol_message';
  connectionId: string;
  data: ProtocolIncomingMessage;
};

export type SnackTransportEventSendMessage = {
  type: 'send_message';
  data: ProtocolOutgoingMessage;
};

export type SnackTransportEventSyntheticMessage = {
  type: 'synthetic_event';
  data: any;
};

export type SnackTransportEvent =
  | SnackTransportEventConnect
  | SnackTransportEventDisconnect
  | SnackTransportEventProtocolMessage
  | SnackTransportEventSendMessage
  | SnackTransportEventSyntheticMessage;

export type SnackTransportListener = (event: SnackTransportEvent) => void;

export interface SnackTransport {
  addEventListener(type: 'message', listener: SnackTransportListener): void;
  removeEventListener(type: 'message', listener: SnackTransportListener): void;
  postMessage(message: SnackTransportMessage): void;
}

export type SnackTransportOptions = {
  channel?: string;
  apiURL?: string;
  verbose?: boolean;
  name?: string;
};
