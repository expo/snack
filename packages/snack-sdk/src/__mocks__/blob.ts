export default class Blob {
  static instances: Blob[] = [];

  config: any;

  constructor(config: any) {
    this.config = config;
    Blob.instances.push(this);
  }
}

// @ts-ignore
global.Blob = Blob;

beforeEach(() => {
  Blob.instances = [];
});
