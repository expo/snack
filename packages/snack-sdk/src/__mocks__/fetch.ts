import type { Server } from 'http';

import { defaultConfig } from '..';
import createSnackager from './snackager';
import createVersions from './versions';
import createWww from './www';

// console.log(defaultConfig.apiURL);

const actualFetch = jest.requireActual('node-fetch');

jest.mock('node-fetch', () => {
  return jest.fn(async (url: string, options: Request) => {
    let server: Server | undefined;

    if (!process.env.SNACK_ENV || process.env.SNACK_ENV === 'mock') {
      // Snackager
      if (url.startsWith(defaultConfig.snackagerURL)) {
        const path = url.substring(defaultConfig.snackagerURL.length);
        server = createSnackager().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;

        // BundledNativeModules.json
      } else if (url.startsWith('https://cdn.jsdelivr.net/npm/expo@')) {
        const path = url.substring('https://cdn.jsdelivr.net/npm'.length);
        server = createVersions().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;

        // www
      } else if (url.startsWith(defaultConfig.apiURL)) {
        const path = url.substring(defaultConfig.apiURL.length);
        server = createWww().listen(0);
        // @ts-ignore: Property port does not exist on type AddressInfo | string
        url = `http://127.0.0.1:${server.address().port}${path}`;
        // console.log('WWW: ', path);
      } else {
        console.log('UNKNOWN: ', url);
      }
    }

    try {
      return await actualFetch(url, options);
    } catch (e) {
      throw e;
    } finally {
      server?.close();
    }
  });
});

const mockFetch = require('node-fetch');

beforeEach(() => mockFetch.mockClear());

export default mockFetch;
