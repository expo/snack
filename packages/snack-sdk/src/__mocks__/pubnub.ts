import {
  PublishParameters,
  PubnubStatus,
  PublishResponse,
  ListenerParameters,
  SubscribeParameters,
  UnsubscribeParameters,
} from 'pubnub';

const TEST_DEVICES: any = {
  ios: {
    platform: 'ios',
    id: '7682-1234',
    name: 'iPhone 8',
  },
  android: {
    platform: 'android',
    id: 'xxxx-xxxx',
    name: 'Nexus 8',
  },
};

export default class PubNub {
  static instances: PubNub[] = [];
  static devices = TEST_DEVICES;

  config: any;
  listener: ListenerParameters = {};
  channel: string = '';
  publishes: PublishParameters[] = [];
  connectIdx = 0;

  constructor(config: any) {
    this.config = config;
    PubNub.instances.push(this);
  }

  addListener = jest.fn((params: ListenerParameters) => {
    this.listener = params;
  });

  removeListener = jest.fn((_params: ListenerParameters) => {
    this.listener = {};
  });

  subscribe = jest.fn((params: SubscribeParameters) => {
    // @ts-expect-error
    this.channel = params.channels[0];
  });

  unsubscribe = jest.fn((_params: UnsubscribeParameters) => {
    this.channel = '';
  });

  stop = jest.fn(() => {
    this.listener = {};
    this.channel = '';
  });

  publish = jest.fn(
    (
      params: PublishParameters,
      _callback: (status: PubnubStatus, response: PublishResponse) => void
    ) => {
      this.publishes.push(params);
      return Promise.resolve();
    }
  );

  connect(uuid?: string): string {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.presence({
      action: 'join',
      uuid: uuid ?? '',
      channel: this.channel,
    });
    return uuid ?? '';
  }

  disconnect(uuid?: string, timeout?: boolean) {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.presence({
      action: timeout ? 'timeout' : 'leave',
      uuid: uuid ?? '',
      channel: this.channel,
    });
    return uuid ?? '';
  }

  sendMessage(message: any, uuid?: string) {
    const device = TEST_DEVICES[uuid ?? 'ios'];
    uuid = device ? JSON.stringify(device) : uuid;
    // @ts-expect-error
    this.listener.message({
      channel: this.channel,
      message,
      publisher: uuid ?? '',
    });
  }
}

beforeEach(() => {
  PubNub.instances = [];
});
