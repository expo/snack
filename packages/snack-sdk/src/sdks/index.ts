import { SDKSpec } from './types';

const assets = {
  // @expo/snack-static/react-native-logo.png is an alias that is
  // implemented in the runtime, and uses by the react-native image docs
  '@expo/snack-static/react-native-logo.png': '*',
};

const unimodules: { [name: string]: '*' } = {
  '@unimodules/core': '*',
  '@unimodules/react-native-adapter': '*',
};
const legacyUnimoduleInterfaces: { [name: string]: '*' } = {
  'unimodules-barcode-scanner-interface': '*',
  'unimodules-camera-interface': '*',
  'unimodules-constants-interface': '*',
  'unimodules-face-detector-interface': '*',
  'unimodules-file-system-interface': '*',
  'unimodules-font-interface': '*',
  'unimodules-image-loader-interface': '*',
  'unimodules-permissions-interface': '*',
  'unimodules-sensors-interface': '*',
  'unimodules-task-manager-interface': '*',
};

const sdks: { [version: string]: SDKSpec } = {
  '39.0.0': {
    version: '^39.0.0',
    coreModules: {
      ...assets,
      ...unimodules,
      ...legacyUnimoduleInterfaces,
      expo: '39.0.5',
      react: '16.13.1',
      'react-native': '0.63.2',
      'react-dom': '*',
      'react-native-web': '*',
    },
    bundledModules: {
      // Packages that require special initialisation (see Modules.tsx)
      'expo-asset': '*',
      'react-native-gesture-handler': '*',
      // 'react-native-vector-icons': '*',
      '@expo/vector-icons': '*',

      // Packages that are used internally by the runtime
      'expo-analytics-amplitude': '*',
      'expo-barcode-scanner': '*',
      'expo-constants': '*',
      'expo-file-system': '*',
      'expo-keep-awake': '*',
      'expo-updates': '*',
      'react-native-view-shot': '*',
      '@react-native-community/async-storage': '*',

      // Packages that are depended on by the "expo" package
      'expo-font': '*',
      'expo-linking': '*',
      'expo-permissions': '*',
      'expo-splash-screen': '*',
      'prop-types': '*',
      'react-native-safe-area-context': '*',

      // Packages that are not really required
      // TODO: THESE SHOULD BE REMOVED ONCE THE SNACK-SDK HANDLES ADDING/REMOVING/UPDATING DEPENDENCIES BETTER
      'expo-camera': '*',
      'expo-gl': '*',
      'expo-image-picker': '*',
      'expo-linear-gradient': '*',
    },
  },
  '40.0.0': {
    version: '^40.0.0',
    coreModules: {
      ...assets,
      ...unimodules,
      ...legacyUnimoduleInterfaces,
      expo: '40.0.1',
      react: '16.13.1',
      'react-native': '0.63.2',
      'react-dom': '*',
      'react-native-web': '*',
    },
    bundledModules: {
      // Packages that require special initialisation (see Modules.tsx)
      'expo-asset': '*',
      'expo-font': '*',
      'react-native-gesture-handler': '*',
      'react-native-safe-area-context': '*',
      // 'react-native-vector-icons': '*',
      '@expo/vector-icons': '*',

      // Packages that are used internally by the runtime
      'expo-constants': '*',
      'expo-file-system': '*',
      'expo-permissions': '*',
      'expo-updates': '*',
      '@react-native-community/async-storage': '*',

      // Common packages that are included for easy of use
      'prop-types': '*',
    },
  },
  '41.0.0': {
    version: '^41.0.0',
    coreModules: {
      ...assets,
      ...unimodules,
      ...legacyUnimoduleInterfaces,
      expo: '41.0.0',
      react: '16.13.1',
      'react-native': '0.63.2',
      'react-dom': '*',
      'react-native-web': '*',
    },
    bundledModules: {
      // Packages that require special initialisation (see Modules.tsx)
      'expo-asset': '*',
      'expo-font': '*',
      'react-native-gesture-handler': '*',
      'react-native-safe-area-context': '*',
      // 'react-native-vector-icons': '*',
      '@expo/vector-icons': '*',

      // Packages that are used internally by the runtime
      'expo-constants': '*',
      'expo-file-system': '*',
      'expo-permissions': '*',
      'expo-updates': '*',
      '@react-native-async-storage/async-storage': '*',

      // Common packages that are included for easy of use
      'prop-types': '*',
    },
    deprecatedModules: {
      '@react-native-community/async-storage':
        'Async Storage has moved to new organization: https://github.com/react-native-async-storage/async-storage',
      'expo-permissions':
        'Use permissions getters and requesters in specific modules instead, such as MediaLibrary.getPermissionsAsync() and MediaLibrary.requestPermissionsAsync().',
    },
  },
  '42.0.0': {
    version: '^42.0.0',
    coreModules: {
      ...assets,
      ...unimodules,
      expo: '42.0.0',
      react: '16.13.1',
      'react-native': '0.63.4',
      'react-dom': '*',
      'react-native-web': '*',
    },
    bundledModules: {
      'expo-asset': '*',
      'expo-font': '*',
      'react-native-gesture-handler': '*',
      'react-native-safe-area-context': '*',
      '@expo/vector-icons': '*',
      'expo-constants': '*',
      'expo-file-system': '*',
      'expo-updates': '*',
      '@react-native-async-storage/async-storage': '*',
      'prop-types': '*',
    },
    deprecatedModules: {
      '@react-native-community/async-storage':
        'Async Storage has moved to new organization: https://github.com/react-native-async-storage/async-storage',
      'expo-permissions':
        'Use permissions getters and requesters in specific modules instead, such as MediaLibrary.getPermissionsAsync() and MediaLibrary.requestPermissionsAsync().',
    },
  },
  '43.0.0': {
    version: '^43.0.0',
    coreModules: {
      ...assets,
      ...unimodules,
      expo: '~43.0.0-beta.3',
      react: '17.0.1',
      'react-native': '0.64.2',
      'react-dom': '*',
      'react-native-web': '*',
    },
    bundledModules: {
      'expo-asset': '*',
      'expo-font': '*',
      'react-native-gesture-handler': '*',
      'react-native-safe-area-context': '*',
      '@expo/vector-icons': '*',
      'expo-constants': '*',
      'expo-file-system': '*',
      'expo-updates': '*',
      '@react-native-async-storage/async-storage': '*',
      'prop-types': '*',
    },
    deprecatedModules: {
      '@react-native-community/async-storage':
        'Async Storage has moved to new organization: https://github.com/react-native-async-storage/async-storage',
      'expo-permissions':
        'Use permissions getters and requesters in specific modules instead, such as MediaLibrary.getPermissionsAsync() and MediaLibrary.requestPermissionsAsync().',
    },
  },
};

export default sdks;
