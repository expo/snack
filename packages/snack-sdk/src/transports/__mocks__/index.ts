import { SnackTransport, SnackTransportOptions } from '../index';
import TestTransport from './TestTransport';

export function createTransport(options: SnackTransportOptions): SnackTransport {
  const result = new TestTransport(options);
  return result;
}

export function createSnackPubTransport(options: SnackTransportOptions): SnackTransport {
  return new TestTransport(options);
}
