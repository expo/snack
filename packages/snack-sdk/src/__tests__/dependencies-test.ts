import '../__mocks__/fetch';
import { SDKVersion } from '../sdks/types';
import Snack, { standardizeDependencies } from './snack-sdk';

// A set of SDK versions to test against.
// When upgrading SDK version, make sure to update this list.
const sdkVersions: { [key: string]: SDKVersion } = {
  prev: '41.0.0',
  current: '42.0.0',
  next: '43.0.0',
};

describe('dependencies', () => {
  it('resolves dependency', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      dependencies: {
        'expo-firebase-analytics': { version: '2.4.1' },
      },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies).toMatchSnapshot();
  });

  it('resolves multiple dependencies', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      dependencies: {
        'expo-font': { version: '8.1.0' },
        '@react-navigation/native': { version: '5.1.1' },
        'react-native-paper': { version: '3.10.1' },
      },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(3);
    expect(state.dependencies).toMatchSnapshot();
  });

  it('ignores resolved dependencies', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      dependencies: {
        'expo-firebase-analytics': {
          version: '8.1.0',
          handle: 'snackager-1/expo-firebase-analytics@8.1.0',
        },
      },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies).toMatchSnapshot();
  });

  it('does not resolve preloaded modules', async () => {
    const snack = new Snack({
      dependencies: {
        'expo-font': {
          version: '8.1.0',
        },
      },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies['expo-font'].handle).toBeUndefined();
  });

  it('adds dependencies', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
    });
    snack.updateDependencies({
      'expo-font': { version: '8.1.0' },
      '@react-navigation/native': { version: '5.1.1' },
      'react-native-paper': { version: '3.10.1' },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(3);
    expect(state.dependencies).toMatchSnapshot();
  });

  it('removes dependencies', async () => {
    const snack = new Snack({});
    snack.updateDependencies({
      'expo-font': { version: '8.1.0' },
      '@react-navigation/native': { version: '5.1.1' },
      'react-native-paper': { version: '3.10.1' },
    });
    snack.updateDependencies({
      'expo-font': null,
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(2);
    expect(state.dependencies).toMatchSnapshot();
  });

  it('loads peer dependencies', async () => {
    const snack = new Snack({});
    snack.updateDependencies({
      firestorter: { version: '2.0.1' },
    });
    const state = await snack.getStateAsync();
    expect(state.dependencies.firestorter.peerDependencies).toMatchSnapshot();
    expect(Object.keys(state.missingDependencies).length).toBe(1);
    expect(state.missingDependencies.mobx).toBeDefined();
    expect(state.missingDependencies).toMatchSnapshot();
  });

  it('fails on invalid dependency name', async () => {
    const snack = new Snack({});
    const name = '+-/.';
    snack.updateDependencies({
      [name]: { version: '1.0.0' },
    });
    const state = snack.getState();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeDefined();
    expect(state.dependencies[name].error).toMatchSnapshot();
  });

  it('fails on incomplete scoped names', async () => {
    const snack = new Snack({});
    const name = '@scopeonly';
    snack.updateDependencies({
      [name]: { version: '1.0.0' },
    });
    const state = snack.getState();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeDefined();
    expect(state.dependencies[name].error).toMatchSnapshot();
  });

  it('succeeds on dependency name with subpath', async () => {
    const snack = new Snack({});
    const name = 'react-native-gesture-handler/DrawerLayout';
    snack.updateDependencies({
      [name]: { version: '1.6.0' },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeUndefined();
    expect(state.dependencies).toMatchSnapshot();
  });

  it('succeeds on scoped dependency name with subpath', async () => {
    const snack = new Snack({});
    const name = '@expo/vector-icons/FontAwesome.ttf';
    snack.updateDependencies({
      [name]: { version: '1.6.0' },
    });
    const state = snack.getState();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeUndefined();
  });

  it('fails on invalid dependency version', async () => {
    const snack = new Snack({});
    const name = 'expo';
    snack.updateDependencies({
      [name]: { version: '-- 2 4' },
    });
    const state = snack.getState();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeDefined();
    expect(state.dependencies[name].error).toMatchSnapshot();
  });

  it('fails on unknown dependency', async () => {
    const snack = new Snack({});
    const name = 'dsfljdslfjkldsjdsjfkldsklfjkldsfjkldsfldsjfkldsklfjdklsdjskl__nevernienooit';
    snack.updateDependencies({
      [name]: { version: '1.0.0' },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeDefined();
  });

  it('retries when clearing error', async () => {
    const name = 'dsfljdslfjkldsjdsjfkldsklfjkldsfjkldsfldsjfkldsklfjdklsdjskl__nevernienooit';
    const snack = new Snack({
      dependencies: {
        [name]: { version: '1.0.0' },
      },
    });
    await snack.getStateAsync();
    snack.updateDependencies({
      [name]: { version: '1.0.0' },
    });
    const state = await snack.getStateAsync();
    expect(Object.keys(state.dependencies).length).toBe(1);
    expect(state.dependencies[name].error).toBeDefined();
  });

  it('resolves * for unlisted package', async () => {
    const snack = new Snack({});
    snack.updateDependencies({
      firestorter: { version: '*' },
    });
    const state = await snack.getStateAsync();
    expect(state.dependencies.firestorter.version).toBe('*');
    expect(state.dependencies.firestorter.wantedVersion).toBeUndefined();
    expect(state.dependencies.firestorter.handle).toBeDefined();
  });

  it('resolves * to wanted version', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
    });
    snack.updateDependencies({
      'expo-constants': { version: '*' },
    });
    const state = await snack.getStateAsync();
    expect(state.dependencies).toMatchSnapshot();
  });

  it('does not resolve * when disabled', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      disabled: true,
    });
    snack.updateDependencies({
      'expo-constants': { version: '*' },
    });
    const state = await snack.getStateAsync();
    expect(state.dependencies).toMatchSnapshot();
  });

  it('resolves * after enabling', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      disabled: true,
    });
    snack.updateDependencies({
      'expo-constants': { version: '*' },
    });
    await snack.getStateAsync();
    snack.setDisabled(false);
    const state = await snack.getStateAsync();
    expect(state.dependencies).toMatchSnapshot();
  });

  it('updates preloaded module version when changing SDK version', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      dependencies: { 'expo-av': { version: '*' } },
    });
    const state1 = await snack.getStateAsync();
    expect(state1.dependencies).toMatchSnapshot();
    snack.setSDKVersion(sdkVersions.next);
    const state2 = await snack.getStateAsync();
    expect(state2.dependencies).toMatchSnapshot();
    expect(state1).not.toMatchObject(state2);
  });

  it('reports missing peer dependencies', async () => {
    const snack = new Snack({
      sdkVersion: sdkVersions.current,
      dependencies: {
        '@react-navigation/stack': {
          handle: 'snackager-1/@react-navigation~stack@5.10.0',
          version: '^5.9.0',
          peerDependencies: {
            'react-native': '*',
            '@react-navigation/native': '5.0.6',
            'react-native-screens': '^2.0.0',
            'react-native-gesture-handler': '^1.0.0',
            'react-native-safe-area-context': '^0.6.0',
            '@react-native-community/masked-view': '^0.1.0',
          },
        },
        '@react-navigation/drawer': {
          handle: 'snackager-1/@react-navigation~drawer@5.10.0',
          version: '^5.9.0',
          peerDependencies: {
            'react-native': '*',
            '@react-navigation/native': '5.0.5',
            'react-native-screens': '^2.0.0',
            'react-native-reanimated': '^1.0.0',
            'react-native-gesture-handler': '^1.0.0',
            'react-native-safe-area-context': '^0.7.0',
          },
        },
      },
    });
    const state = await snack.getStateAsync();
    expect(state.missingDependencies).toMatchSnapshot();
  });
});

describe('standardizeDependencies', () => {
  it('converts v1 dependencies', () => {
    expect(
      standardizeDependencies({
        dep1: '1.2.3',
      })
    ).toMatchObject({
      dep1: {
        version: '1.2.3',
      },
    });
  });
  it('converts v2 dependencies', () => {
    expect(
      standardizeDependencies({
        dep1: {
          version: '1.2.3',
          peerDependencies: {
            peerDep2: {
              version: '4.5.6',
            },
          },
        },
      })
    ).toMatchObject({
      dep1: {
        version: '1.2.3',
        peerDependencies: {
          peerDep2: '4.5.6',
        },
      },
    });
  });
  it('returns v3 dependencies untouched', () => {
    const deps = {
      dep1: {
        version: '1.2.3',
        peerDependencies: {
          peerDep2: '4.5.6',
        },
      },
    };
    expect(standardizeDependencies(deps)).toBe(deps);
  });
});
