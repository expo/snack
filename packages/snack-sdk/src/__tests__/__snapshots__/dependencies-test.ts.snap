// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependencies adds dependencies 1`] = `
Object {
  "@react-navigation/native": Object {
    "handle": "snackager-1/@react-navigation~native@5.1.1",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "5.1.1",
  },
  "expo-font": Object {
    "version": "8.1.0",
    "wantedVersion": "~9.2.1",
  },
  "react-native-paper": Object {
    "handle": "snackager-1/react-native-paper@3.10.1",
    "peerDependencies": Object {
      "@expo/vector-icons": null,
      "react": "*",
      "react-native": "*",
    },
    "version": "3.10.1",
  },
}
`;

exports[`dependencies does not resolve * when disabled 1`] = `
Object {
  "expo-constants": Object {
    "version": "*",
  },
}
`;

exports[`dependencies fails on incomplete scoped names 1`] = `[Error: Invalid dependency '@scopeonly' (name can only contain URL-friendly characters)]`;

exports[`dependencies fails on invalid dependency name 1`] = `[Error: Invalid dependency '+-/.' (name can only contain URL-friendly characters)]`;

exports[`dependencies fails on invalid dependency version 1`] = `[Error: Invalid dependency 'expo' (version '-- 2 4' is not a valid semver)]`;

exports[`dependencies ignores resolved dependencies 1`] = `
Object {
  "expo-firebase-analytics": Object {
    "handle": "snackager-1/expo-firebase-analytics@8.1.0",
    "version": "8.1.0",
    "wantedVersion": "~4.1.0",
  },
}
`;

exports[`dependencies loads peer dependencies 1`] = `
Object {
  "mobx": "^4.0.0",
}
`;

exports[`dependencies loads peer dependencies 2`] = `
Object {
  "mobx": Object {
    "dependents": Array [
      "firestorter",
    ],
    "wantedVersion": "^4.0.0",
  },
}
`;

exports[`dependencies removes dependencies 1`] = `
Object {
  "@react-navigation/native": Object {
    "handle": "snackager-1/@react-navigation~native@5.1.1",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "5.1.1",
  },
  "react-native-paper": Object {
    "handle": "snackager-1/react-native-paper@3.10.1",
    "peerDependencies": Object {
      "@expo/vector-icons": null,
      "react": "*",
      "react-native": "*",
    },
    "version": "3.10.1",
  },
}
`;

exports[`dependencies reports missing peer dependencies 1`] = `
Object {
  "@react-native-community/masked-view": Object {
    "dependents": Array [
      "@react-navigation/stack",
    ],
    "wantedVersion": "0.1.10",
  },
  "@react-navigation/native": Object {
    "dependents": Array [
      "@react-navigation/stack",
      "@react-navigation/drawer",
    ],
    "wantedVersion": "5.0.6",
  },
  "react-native-gesture-handler": Object {
    "dependents": Array [
      "@react-navigation/stack",
      "@react-navigation/drawer",
    ],
    "wantedVersion": "~1.10.2",
  },
  "react-native-reanimated": Object {
    "dependents": Array [
      "@react-navigation/drawer",
    ],
    "wantedVersion": "~2.2.0",
  },
  "react-native-safe-area-context": Object {
    "dependents": Array [
      "@react-navigation/stack",
      "@react-navigation/drawer",
    ],
    "wantedVersion": "3.2.0",
  },
  "react-native-screens": Object {
    "dependents": Array [
      "@react-navigation/stack",
      "@react-navigation/drawer",
    ],
    "wantedVersion": "~3.4.0",
  },
}
`;

exports[`dependencies resolves * after enabling 1`] = `
Object {
  "expo-constants": Object {
    "version": "~11.0.1",
    "wantedVersion": "~11.0.1",
  },
}
`;

exports[`dependencies resolves * to wanted version 1`] = `
Object {
  "expo-constants": Object {
    "version": "~11.0.1",
    "wantedVersion": "~11.0.1",
  },
}
`;

exports[`dependencies resolves dependency 1`] = `
Object {
  "expo-firebase-analytics": Object {
    "handle": "snackager-1/expo-firebase-analytics@2.4.1",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "2.4.1",
    "wantedVersion": "~4.1.0",
  },
}
`;

exports[`dependencies resolves multiple dependencies 1`] = `
Object {
  "@react-navigation/native": Object {
    "handle": "snackager-1/@react-navigation~native@5.1.1",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "5.1.1",
  },
  "expo-font": Object {
    "version": "8.1.0",
    "wantedVersion": "~9.2.1",
  },
  "react-native-paper": Object {
    "handle": "snackager-1/react-native-paper@3.10.1",
    "peerDependencies": Object {
      "@expo/vector-icons": null,
      "react": "*",
      "react-native": "*",
    },
    "version": "3.10.1",
  },
}
`;

exports[`dependencies succeeds on dependency name with subpath 1`] = `
Object {
  "react-native-gesture-handler/DrawerLayout": Object {
    "handle": "snackager-1/react-native-gesture-handler~DrawerLayout@1.6.0",
    "peerDependencies": Object {
      "react-native-gesture-handler": "1.6.0",
    },
    "version": "1.6.0",
    "wantedVersion": "~1.10.2",
  },
}
`;

exports[`dependencies updates preloaded module version when changing SDK version 1`] = `
Object {
  "expo-av": Object {
    "handle": "snackager-1/expo-av@9.2.3",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "~9.2.3",
    "wantedVersion": "~9.2.3",
  },
}
`;

exports[`dependencies updates preloaded module version when changing SDK version 2`] = `
Object {
  "expo-av": Object {
    "handle": "snackager-1/expo-av@10.1.0",
    "peerDependencies": Object {
      "react": "*",
      "react-native": "*",
    },
    "version": "~10.1.0",
    "wantedVersion": "~10.1.0",
  },
}
`;
